wgpu-types/src/lib.rs

```diff
@@ -14,6 +14,48 @@ use serde::{Deserialize, Serialize};
 use std::hash::{Hash, Hasher};
 use std::{num::NonZeroU32, ops::Range};
 
+// Use this macro instead of the one provided by the bitflags_serde_shim crate
+// because the latter produces an error when deserializing bits that are not
+// specified in the bitflags, while we want deserialization to succeed and
+// and unspecified bits to lead to errors handled in wgpu-core.
+// Note that plainly deriving Serialize and Deserialized would have a similar
+// behavior to this macro (unspecified bit do not produce an error).
+macro_rules! impl_bitflags {
+    ($name:ident) => {
+        #[cfg(feature = "trace")]
+        impl serde::Serialize for $name {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                self.bits().serialize(serializer)
+            }
+        }
+
+        #[cfg(feature = "replay")]
+        impl<'de> serde::Deserialize<'de> for $name {
+            fn deserialize<D>(deserializer: D) -> Result<$name, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                let value = <_ as serde::Deserialize<'de>>::deserialize(deserializer)?;
+                // Note: newer version of bitflags replaced from_bits_unchecked with
+                // from_bits_retain which is not marked as unsafe (same implementation).
+                Ok(unsafe { $name::from_bits_unchecked(value) })
+            }
+        }
+
+        impl $name {
+            /// Returns true if the bitflags contains bits that are not part of
+            /// the bitflags definition.
+            pub fn contains_invalid_bits(&self) -> bool {
+                let all = Self::all().bits();
+                self.bits() | all != all
```

Comment:
This has _very_ confusing order of operations, I stared at this for like a minute to get what it's doing :)

Reply:
Is it confusing because the negation? I read it as "the union of the real bits and the set of know bits is different from the set of known bits" or in other words "the set of real bits is not contained in the set of known bits".

I could rewrite this as "There are bits in the set of unknown bits", aka:  `self.bits() & !all != 0` which felt a bit more complicated to me because there are two negations instead of one but but everyone has their own way oflaying out this kind of logic in their head. Let me know what you prefer (or if you have a different formulation in mind).

Reply:
All I found confusing was the literal precidence of the operation. I kept trying to read it as `self.bits() | (all != all)` which obviously makes no sense. So just adding parens (or making the left side a variable) would clear it up for me.

Reply:
Oh, I see, makes sense. Done.